neqo-http3/src/send_message.rs

```diff
@@ -281,29 +199,19 @@ impl SendStream for SendMessage {
 }
 
 impl HttpSendStream for SendMessage {
-    fn set_message(
-        &mut self,
-        headers: &[Header],
-        data: Option<&[u8]>,
-        conn: &mut Connection,
-    ) -> Res<()> {
-        if !matches!(self.state, SendMessageState::Uninitialized) {
-            return Err(Error::AlreadyInitialized);
-        }
+    fn send_headers(&mut self, headers: &[Header], conn: &mut Connection) {
```

Comment:
This isn't doing anything to ensure that headers are injected at a sensible position in the stream.  I think that we want to do that so that we don't do allow people using this code to do silly things, either on purpose or by accident.

That is, we should validate that:

* there aren't multiple requests (with methods and so forth)
* there aren't multiple final status codes
* informational status codes don't follow final status codes
* requests and response blocks (those with methods or status codes) don't follow data
* there is just one block of trailers
* data doesn't follow trailers

That suggests a state machine of sorts.  There might be a point to specializing into RequestSendStream and ResponseSendStream to do that, but I'm not sure if that really helps.

I'm happy for that to exist at a higher level than this class or in a specialization of this class, but as we're basically exposing this function directly, we should put some safety rails on it.

Reply:
I purposely decided not to do that. The main reason was to easy testing the client side, by letting  server do silly things. I did not want to have asymmetry, i.e. client is strict and sever is not. The previous version had states that are appropriate for the client side(the states fit well the Firefox use case), but not really working for the server. 

I will try to implement checks and keep one SendMessage and if it too ugly I will split them.